### PR TITLE
Update type definitions with rector

### PR DESCRIPTION
**Pull request by thePanz**
- [x] Update types definitions with Rector PHP
- [x] Fix phpstan outdated configuration
- [x] Upgrade Phpstan analysis to level 7
- [x] Update php-cs-fixer settings (use shared cache location)

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -1,6 +1,7 @@
 .gitattributes export-ignore
 .gitignore export-ignore
-.php_cs.dist export-ignore
+rector.php
```
Comment (mjanser) [inline .gitattributes]: Should this be `export-ignore` as well?

**Review by mjanser — APPROVED** [review]

File: .gitattributes (hunk 1 above)
Comment (dbu) [reply .gitattributes]: imo yes :-)

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -26,12 +26,12 @@ jobs:
...
           php-version: ${{ matrix.php-version }}
       - name: Remove dev tools to not interfere with dependencies
-        run: composer remove --dev friendsofphp/php-cs-fixer phpstan/phpstan-phpunit phpstan/phpstan
+        run: composer remove --dev friendsofphp/php-cs-fixer phpstan/phpstan-phpunit phpstan/phpstan rector/rector
       - name: Composer cache
         uses: actions/cache@v3
         with:
           path: ${{ env.HOME }}/.composer/cache
-          key: ${{ runner.os }}-php-${{ hashFiles('**/composer.lock') }}
+          key: ${{ runner.os }}-php-${{ hashFiles('**/composer.json') }}
```
Comment (dbu) [inline .github/workflows/ci.yml:34]: oh. we don't even have a checked in lock file, right?

File: src/Configuration/ClassToGenerate.php [hunk 3]
```diff
@@ -4,42 +4,38 @@
...
-    private $defaultVersions;
+    private ?array $defaultVersions;
 
-    public function __construct(GeneratorConfiguration $configuration, string $className, array $defaultVersions = null)
-    {
-        $this->configuration = $configuration;
-        $this->className = $className;
+    /**
+     * @param class-string      $className       Fully qualified class name
+     * @param list<string>|null $defaultVersions
```
Comment (dbu) [inline src/Configuration/ClassToGenerate.php]: should these be here or as var annotations on the promoted arguments?

File: src/Configuration/GroupCombination.php [hunk 4]
```diff
@@ -7,41 +7,32 @@
...
-    {
-        $this->containingClass = $containingClass;
-        $this->groups = $groups;
+    public function __construct(
+        private ClassToGenerate $containingClass,
+        /**
+         * One combination of groups to generate.
+         * An empty array means to generate with no groups.
+         *  e.g. ['api', 'details'].
+         */
```
Comment (dbu) [inline src/Configuration/GroupCombination.php:18]: here we have the comment on the promoted argument and not as `@param` on the constructor method

**Review by dbu — APPROVED** [review]
thanks for the maintenance work!

File: .github/workflows/ci.yml (hunk 2 above)
Comment (thePanz) [reply .github/workflows/ci.yml:34]: exactly, that's why the cache was not properly used yet :)

File: src/Configuration/GroupCombination.php (hunk 4 above)
Comment (thePanz) [reply src/Configuration/GroupCombination.php:18]: this constructor is for "all promoted" arguments, the first one is mixing promoted and non-promoted ones

I also would like to use the same pattern, but this conflicts with php-cs-fixer :(

**Pull request merged**